New build tool.

hugelgupf opened the pull request:
Provides a new build tool to replace bb/bb.go and scripts/ramfs.go with a unified interface.

 - Reduces lzma-compressed bb-initramfs size by 1M to 2.8M :) (we didn't used to strip symbols, courtesy of copy and pasting go build commands all over the place...)

prattmic commented on pkg/golang/build.go:
```diff
@@ -0,0 +1,129 @@
+package golang
```
// Package golang ...

prattmic replied:
There are lots of missing comments in this package.

prattmic commented on pkg/golang/build.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+// FindPackage returns the full path to `pkg` according to the context's Gopaths.
+func (c Environ) FindPackage(pkg string) (string, error) {
```
Wouldn't `types.ImporterFrom.ImportFrom` be more robust?

https://godoc.org/go/types#ImporterFrom

I think you're supposed to get it with `importer.For("source", nil).(types.ImporterFrom)`? bleh.

https://godoc.org/go/importer#For

prattmic replied:
Argh, I see this is already done in a later commit.

prattmic commented on pkg/golang/build.go:
```diff
@@ -0,0 +1,129 @@
+		"build",
+		"-a", // Force rebuilding of packages.
+		"-o", binaryPath,
+		"-installsuffix", "cgo",
```
Even for no cgo?

hugelgupf commented:
There's a working, really gross bb mode here now.

rminnich commented:
This broke NiChrome but only because of the globbing change, I'll just work around it. It's too much of an improvement not tohave. You've tested with dynamic compilation in a VM or something I assume?

hugelgupf commented:
source mode (scripts/ramfs.go equivalent) is still broken, but bb mode is good to go. Also rewrote the README and added some (but not enough) documentation to Go source.

hugelgupf commented:
Source mode now works, too.
